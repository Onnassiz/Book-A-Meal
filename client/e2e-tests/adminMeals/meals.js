/* eslint func-names: 0, no-unused-expressions: 0 */
const path = require('path');

const imagePath = path.join(__dirname, '../../assets/images/3.jpg');

const meals = {
  'Should sign user into the application': (client) => {
    client
      .url('http://localhost:3002/#/')
      .waitForElementVisible('header')
      .moveToElement('#login', 10, 10)
      .click('#login')
      .waitForElementVisible('h3')
      .assert.containsText('h3', 'Sign In')
      .setValue('#signInEmail', 'onnassiz@gmail.com')
      .setValue('#signInPassword', 'password')
      .click('input[type=submit]')
      .waitForElementVisible('header')
      .assert.visible('#caterer_dropdown')
      .assert.visible('#user_dropdown');
  },

  'Should navigate to the admin meals page': (client) => {
    client
      .moveToElement('#caterer_dropdown', 10, 10)
      .pause(1000)
      .waitForElementVisible('#admin_meals')
      .assert.visible('#admin_meals')
      .click('#admin_meals')
      .waitForElementVisible('#addMeal')
      .pause(2000)
      .assert.visible('#addMeal');
  },

  'Should show new meal modal form': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .pause(1000)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .pause(1000)
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(2000);
  },

  'Should fill meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Fried Rice')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form again': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill another meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Chicken')
      .setValue('#price', 2000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal. It has all you need. Try it out.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show a new meal modal form for the third time': (client) => {
    client
      .moveToElement('#addMeal', 10, 10)
      .assert.containsText('#addMeal', 'Add Meal')
      .click('#addMeal')
      .waitForElementVisible('.box')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should fill the third meal form and create a new meal': (client) => {
    client
      .pause(1000)
      .setValue('#name', 'Beans')
      .setValue('#price', 1000)
      .setValue('#category', 'Dinner')
      .setValue('#description', 'This is a good meal for me and you.')
      .moveToElement('input[type=submit]', 10, 10)
      .pause(2000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show an update meal modal': (client) => {
    client
      .waitForElementVisible('#updateModal:first-of-type')
      .assert.visible('#updateModal:first-of-type')
      .click('#updateModal:first-of-type')
      .assert.visible('.box')
      .pause(1000);
  },

  'Should update the third meal form and update a new meal': (client) => {
    client
      .pause(1000)
      .clearValue('#name')
      .setValue('#name', 'Rice')
      .clearValue('#price')
      .setValue('#price', 1200)
      .moveToElement('input[type=submit]', 10, 10)
      .pause(1000)
      .click('input[type=submit]')
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 5000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default');
  },

  'Should show delete modal and delete meal on button click': (client) => {
    client
      .assert.visible('#deleteModal:first-of-type')
      .click('#deleteModal:first-of-type')
      .waitForElementVisible('button.swal-button.swal-button--confirm.swal-button--danger', 10000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm.swal-button--danger')
      .waitForElementVisible('button.swal-button.swal-button--confirm', 15000)
      .pause(1000)
      .click('button.swal-button.swal-button--confirm')
      .waitForElementNotVisible('button.swal-button.swal-button--confirm', 20000)
      .assert.hidden('button.swal-button.swal-button--confirm')
      .pause(1000);
  },

  'Should show an image upload modal modal': (client) => {
    client
      .assert.visible('#addPhotoModal:first-of-type')
      .click('#addPhotoModal:first-of-type')
      .waitForElementVisible('#dropzoneHeader')
      .assert.containsText('#dropzoneHeader', 'Drop Image Here')
      .pause(1000)
      .moveToElement('#upload_image', 10, 10)
      .pause(1000)
      .setValue('input[type=file]', imagePath)
      .waitForElementVisible('div.Toastify__toast.Toastify__toast--default', 15000)
      .assert.visible('div.Toastify__toast.Toastify__toast--default')
      .pause(1000);
    client.end();
  },
};

export default meals;
