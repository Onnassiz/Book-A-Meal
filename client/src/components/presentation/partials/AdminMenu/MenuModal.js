import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import Pagination from 'react-js-pagination';
import empty from 'is-empty';
import { toast } from 'react-toastify';
import { addMenuModalStyle } from './../../../../utilities/modalStyles';
import { Checkbox, BasicInput } from './../../form/BasicInput';
import ShowErrors from './../../partials/ShowErrors';
import SubmitButton from '../../../presentation/form/SubmitButton';
import validateMenu from '../../../../utilities/validateMenuForm';
import { handleSubmit } from '../../../../utilities/adminMenusHelper';

class MenuModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      closeModalStyle: { float: 'right' },
      errors: {},
      name: '',
      endDate: '',
      selectedMeals: [],
      startDate: new Date().toISOString().split('T')[0],
    };
    this.onChange = this.onChange.bind(this);
    this.isSelected = this.isSelected.bind(this);
    this.onCheckBoxChange = this.onCheckBoxChange.bind(this);
    this.afterOpenModal = this.afterOpenModal.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  onCheckBoxChange(e, meal) {
    const meals = this.state.selectedMeals;
    if (e.target.checked) {
      meals.push({ mealId: meal.id, price: meal.price });
      this.setState({ selectedMeals: meals });
    } else {
      const index = meals.findIndex(x => x.mealId === e.target.value);
      meals.splice(index, 1);
      this.setState({ selectedMeals: meals });
    }
  }

  getMenuExtraDays() {
    const endDate = (new Date(this.state.endDate).getTime() / 1000);
    const startDate = (new Date(this.state.startDate).getTime() / 1000);
    const unixTimeDiff = endDate - startDate;
    return unixTimeDiff > 0 ? unixTimeDiff / 86400 : 0;
  }

  isSelected(id) {
    return this.state.selectedMeals.findIndex(x => x.mealId === id) !== -1;
  }

  afterOpenModal() {
    const { state } = this.props;
    this.resetFields();
    if (state.updateMode) {
      this.setState({
        selectedMeals: state.selectedMeals,
        startDate: state.currentMenu.date,
        name: state.currentMenu.name,
      });
    }
  }

  isValid(formData) {
    this.setState({ errors: {} });
    const { errors, isValid } = validateMenu(formData);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  resetFields() {
    this.setState({
      startDate: '', endDate: '', name: '', selectedMeals: [], updateMode: false,
    });
  }

  createNewMenu(formData) {
    const { postMenu, addToViewMenus } = this.props;
    postMenu(formData).then((response) => {
      if (response.status === 201) {
        toast('Menu(s) successfully created');
        addToViewMenus(response.data.menus);
      }
    });
  }

  updateExistingMenu(formData) {
    const { updateMenu, updateViewMenus } = this.props;
    updateMenu(formData).then((response) => {
      if (response.status === 200) {
        toast('Menu successfully updated');
        updateViewMenus(response.data.menu);
      }
    });
  }

  renderMenuModal() {
    const { state, toggleShowModal } = this.props;
    return (
      <Modal
        isOpen={state.isShowingModal}
        closeTimeoutMS={1}
        onAfterOpen={this.afterOpenModal}
        shouldCloseOnEsc
        shouldCloseOnOverlayClick
        onRequestClose={toggleShowModal}
        style={addMenuModalStyle}
        ariaHideApp={false}
        contentLabel="Modal"
      >
        <div className="col-12">
          <a title="close" onClick={toggleShowModal} style={this.state.closeModalStyle}><i style={{ fontSize: 25 }} className="material-icons">close</i></a>
        </div>
        {empty(state.meals) ?
          <div>
            No meals found. You must add meals to be able to create a Menu.
          </div> : this.renderMenuModalBody()}
      </Modal>
    );
  }

  renderStartDate() {
    const today = new Date().toISOString().split('T')[0];
    return (
      <div>
        <BasicInput
          name="startDate"
          type="date"
          min={today}
          label="Start Date"
          value={this.state.startDate}
          onChange={this.onChange}
          hasError={this.state.errors.date !== undefined}
        />
      </div>
    );
  }

  renderDatesAndName() {
    const today = new Date().toISOString().split('T')[0];
    const { state } = this.props;
    return (
      <div>
        {state.updateMode ? '' :
        <BasicInput
          name="endDate"
          onChange={this.onChange}
          type="date"
          min={this.state.startDate === '' ? today : this.state.startDate}
          label="End Date (optional)"
          value={this.state.endDate}
          hasError={this.state.errors.endDate !== undefined}
        />
        }
        <BasicInput
          name="name"
          type="text"
          label="Menu Name (optional)"
          value={this.state.name}
          onChange={this.onChange}
          hasError={this.state.errors.name !== undefined}
        />
      </div>
    );
  }

  renderMeals() {
    const { state } = this.props;
    return (
      <div id="meals-checkBox">
        {state.meals.map(item =>
          (<Checkbox
            isChecked={this.isSelected(item.id)}
            meal={item}
            key={item.id}
            onChange={e => this.onCheckBoxChange(e, item)}
          />))}
        {this.renderMealsPagination()}
      </div>
    );
  }

  renderMealsPagination() {
    const { state, handlePageChange } = this.props;
    return (
      <div style={{ textAlign: 'center' }}>{state.mealsCount > 9 ?
        <Pagination
          itemsCountPerPage={9}
          totalItemsCount={state.mealsCount}
          pageRangeDisplayed={5}
          hideDisabled
          activePage={state.mealsActivePage}
          onChange={number => handlePageChange(number)}
        /> : ''}
      </div>
    );
  }

  renderMenuModalBody() {
    const { menus, formState, state } = this.props;
    return (
      <div>
        <form onSubmit={e => handleSubmit(e, this)}>
          <div className="box-menu">
            <h3>{this.state.updateMode ? 'Update' : 'Set Daily'} Menu</h3>
            <ShowErrors clientErrors={this.state.errors} serverErrors={menus.errors || {}} />
            {this.renderStartDate()}
            {this.renderDatesAndName()}
            <div style={{ fontSize: 18, marginTop: 45 }}>
              <span className="number-badge">{this.state.selectedMeals.length}</span> meals selected
            </div>
          </div>
          {this.renderMeals()}
          <div id="meals" style={{ textAlign: 'right' }}>
            <hr />
            <SubmitButton className="button" value={state.updateMode ? 'Update' : 'Submit'} isLoading={formState.isLoading} />
          </div>
        </form>
      </div>
    );
  }

  render() {
    return (
      <div>{this.renderMenuModal()}</div>
    );
  }
}

MenuModal.propTypes = {
  state: PropTypes.object.isRequired,
  toggleShowModal: PropTypes.func.isRequired,
  menus: PropTypes.object.isRequired,
  formState: PropTypes.object.isRequired,
  postMenu: PropTypes.func.isRequired,
  updateMenu: PropTypes.func.isRequired,
  addToViewMenus: PropTypes.func.isRequired,
  updateViewMenus: PropTypes.func.isRequired,
  handlePageChange: PropTypes.func.isRequired,
};

export default MenuModal;
