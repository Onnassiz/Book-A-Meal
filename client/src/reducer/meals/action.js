import axios from 'axios';
import { store } from '../../store';
import { setLoading, unsetLoading } from '../formState/action';

export const SET_MEAL = 'SET_MEAL';
export const SET_MEAL_SERVER_ERRORS = 'SET_MEAL_SERVER_ERRORS';
export const SET_MEAL_ALERT = 'SET_MEAL_ALERT';

function setAlert(alert) {
	return dispatch => dispatch({
		type: SET_MEAL_ALERT,
		alert,
	});
}

export function setMeal(meal) {
	let { meals } = store.getState().meals;
	meals = [meal].concat(meals);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function setMealArray(meal) {
	let { meals } = store.getState().meals;
	meals = meals.concat(meal);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function updateMealState(meal) {
	const { meals } = store.getState().meals;
	const index = meals.findIndex(x => x.id === meal.id);
	meals[index] = meal;
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

export function updateMealAfterDelete(id) {
	let { meals } = store.getState().meals;
	meals = meals.filter(x => x.id !== id);
	return dispatch => dispatch({
		type: SET_MEAL,
		meals,
	});
}

function setMealErrors(errors) {
	return dispatch => dispatch({
		type: SET_MEAL_SERVER_ERRORS,
		errors,
	});
}

export function postMeal(meal) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.post('api/v1/meals', meal).then((response) => {
			dispatch(unsetLoading());
			dispatch(setMeal(response.data));
			dispatch(setAlert('Meal successfully created'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response));
			return error;
		});
	};
}

export function updateMeal(meal) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.put(`api/v1/meals/${meal.id}`, meal).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealState(response.data));
			dispatch(setAlert('Meal successfully updated'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}

export function deleteMealById(id) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.delete(`api/v1/meals/${id}`).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealAfterDelete(id));
			dispatch(setAlert('Meal successfully deleted'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}

export function putImage(id, field) {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.put(`api/v1/meals/image/${id}`, field).then((response) => {
			dispatch(unsetLoading());
			dispatch(updateMealState(response.data));
			dispatch(setAlert('Image successfully uploaded'));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			dispatch(setMealErrors(error.response.data));
			return error;
		});
	};
}
export function getMeals() {
	return (dispatch) => {
		dispatch(setLoading());
		return axios.get('api/v1/meals').then((response) => {
			dispatch(unsetLoading());
			dispatch(setMealArray(response.data));
			return response;
		}).catch((error) => {
			dispatch(unsetLoading());
			return error;
		});
	};
}
