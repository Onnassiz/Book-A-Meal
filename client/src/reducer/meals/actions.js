import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_MEAL = 'SET_MEAL';
export const SET_MEAL_SERVER_ERRORS = 'SET_MEAL_SERVER_ERRORS';
export const SET_MEAL_ALERT = 'SET_MEAL_ALERT';

function setAlert(alert) {
  return dispatch => dispatch({
    type: SET_MEAL_ALERT,
    alert,
  });
}

// function setMeal(storeMeals, meal) {
//   const meals = [meal].concat(storeMeals);
//   return dispatch => dispatch({
//     type: SET_MEAL,
//     meals,
//   });
// }

// function setMealArray(storeMeals, meal) {
//   const meals = storeMeals.concat(meal);
//   return dispatch => dispatch({
//     type: SET_MEAL,
//     meals,
//   });
// }

export function updateMealState(storeMeals, meal) {
  const index = storeMeals.findIndex(x => x.id === meal.id);
  const meals = storeMeals;
  meals[index] = meal;
  return dispatch => dispatch({
    type: SET_MEAL,
    meals,
  });
}

export function updateMealAfterDelete(storeMeals, id) {
  const meals = storeMeals.filter(x => x.id !== id);
  return dispatch => dispatch({
    type: SET_MEAL,
    meals,
  });
}

function setMealErrors(errors) {
  return dispatch => dispatch({
    type: SET_MEAL_SERVER_ERRORS,
    errors,
  });
}

export function postMeal(meal) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.post('api/v1/meals', meal).then((response) => {
      dispatch(unsetLoading());
      // dispatch(setMeal(getState().meals.meals, response.data));
      dispatch(setAlert('Meal successfully created'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function updateMeal(meal) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/meals/${meal.id}`, meal).then((response) => {
      dispatch(unsetLoading());
      // dispatch(updateMealState(getState().meals.meals, response.data));
      dispatch(setAlert('Meal successfully updated'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function deleteMealById(id) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.delete(`api/v1/meals/${id}`).then((response) => {
      dispatch(unsetLoading());
      // dispatch(updateMealAfterDelete(getState().meals.meals, id));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setMealErrors(error.response.data));
      return error;
    });
  };
}

export function getMeals(limit = 10, offset = 0) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.get(`api/v1/meals/?limit=${limit}&offset=${offset}`).then((response) => {
      dispatch(unsetLoading());
      // dispatch(setMealArray(getState().meals.meals, response.data));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      return error;
    });
  };
}
