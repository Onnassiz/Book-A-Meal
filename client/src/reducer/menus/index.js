import { SET_MENUS, SET_MENUS_ALERT, SET_MENUS_SERVER_ERRORS, SET_USER_MENUS } from './action';
import { getCurrentDate } from '../../utilities/functions';

const initialState = {
	menus: [],
	userMenus: [],
	currentDate: getCurrentDate(),
};

export default (state = initialState, action) => {
	switch (action.type) {
	case SET_MENUS:
		return { ...state, menus: action.menus };
	case SET_USER_MENUS:
		return { ...state, userMenus: action.menus, currentDate: action.date };
	case SET_MENUS_SERVER_ERRORS:
		return { ...state, errors: action.errors.data };
	case SET_MENUS_ALERT:
		return { ...state, alert: action.alert };
	default:
		return state;
	}
};
