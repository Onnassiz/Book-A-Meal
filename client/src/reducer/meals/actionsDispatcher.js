import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actionsDispatcher';
import { setMealErrorsAction } from './actions';

const setMealErrors = errors => dispatch => dispatch(setMealErrorsAction(errors));

export const postMeal = meal => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/meals', meal).then((response) => {
    dispatch(unsetLoading());
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setMealErrors(error.response.data));
    return error;
  });
};

export const updateMeal = meal => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/meals/${meal.id}`, meal).then((response) => {
    dispatch(unsetLoading());
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setMealErrors(error.response.data));
    return error;
  });
};

export const deleteMealById = id => (dispatch) => {
  dispatch(setLoading());
  return axios.delete(`api/v1/meals/${id}`).then((response) => {
    dispatch(unsetLoading());
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setMealErrors(error.response.data));
    return error;
  });
};

export const getMeals = (limit = 10, offset = 0) => (dispatch) => {
  dispatch(setLoading());
  return axios.get(`api/v1/meals/?limit=${limit}&offset=${offset}`).then((response) => {
    dispatch(unsetLoading());
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setMealErrors(error.response.data));
    return error;
  });
};
