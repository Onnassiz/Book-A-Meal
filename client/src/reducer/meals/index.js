import { SET_MEAL, SET_MEAL_SERVER_ERRORS } from './actions';

const initialState = {
  meals: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_MEAL:
      return { ...state, meals: action.meals };
    case SET_MEAL_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
