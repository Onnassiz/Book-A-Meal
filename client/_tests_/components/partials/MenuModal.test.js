import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MenuModal from '../../../src/components/presentation/partials/AdminMenu/MenuModal';
import { getMealsMock, singleMeal } from '../../objectProps/meals.props';
import { singleMenu } from '../../objectProps/menus.props';
import { postMenu, updateMenu } from '../../promiseProps/menus.props';

const props = {
  state: {
    selectedMeals: [], updateMode: true, meals: getMealsMock(5), currentMenu: singleMenu,
  },
  toggleShowModal: jest.fn(),
  menus: { errors: {} },
  formState: { isLoading: false },
  postMenu,
  updateMenu,
  addToViewMenus: jest.fn(),
  updateViewMenus: jest.fn(),
  handlePageChange: jest.fn(),
};

const wrapper = shallow(<MenuModal {...props} />);

describe('<MenuModal />', () => {
  it('Should render Menu Modal page successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should call onChange method to update handle input change', () => {
    const onChangeSpy = jest.spyOn(wrapper.instance(), 'onChange');
    const firstEvent = {
      target: {
        name: 'startDate',
        value: '2019-12-12',
      },
    };
    wrapper.instance().onChange(firstEvent);
    expect(wrapper.state().startDate).toBe('2019-12-12');
    expect(onChangeSpy).toHaveBeenCalled();
  });

  it('Should call onCheckBoxChange method to update handle input change', () => {
    const onCheckBoxChangeSpy = jest.spyOn(wrapper.instance(), 'onCheckBoxChange');
    const event = {
      target: {
        checked: true,
      },
    };
    const meal = singleMeal;
    wrapper.instance().onCheckBoxChange(event, meal);

    expect(wrapper.state().selectedMeals.length).toBe(1);
    expect(onCheckBoxChangeSpy).toHaveBeenCalled();
  });

  it('Should call onCheckBoxChange method to update handle input change when checkbox is unchecked', () => {
    const onCheckBoxChangeSpy = jest.spyOn(wrapper.instance(), 'onCheckBoxChange');
    const event = {
      target: {
        checked: false,
      },
    };

    const meal = singleMeal;
    wrapper.instance().onCheckBoxChange(event, meal);

    expect(wrapper.state().selectedMeals.length).toBe(0);
    expect(onCheckBoxChangeSpy).toHaveBeenCalled();
  });

  it('Should call isSelected method and return a boolean value that tells if a meal is checked', () => {
    const isSelectedSpy = jest.spyOn(wrapper.instance(), 'isSelected');

    wrapper.setState({ selectedMeals: [{ mealId: singleMeal.id }] });

    expect(wrapper.instance().isSelected('wrong')).toBe(false);
    expect(wrapper.instance().isSelected(singleMeal.id)).toBe(true);
    expect(isSelectedSpy).toHaveBeenCalled();
  });

  it('Should call afterOpenModal and set modal fields when modal pops up', () => {
    const afterOpenModalSpy = jest.spyOn(wrapper.instance(), 'afterOpenModal');

    wrapper.setState({ selectedMeals: [{ mealId: singleMeal.id }] });

    wrapper.instance().afterOpenModal();

    expect(wrapper.state().name).toBe(singleMenu.name);
    expect(wrapper.state().startDate).toBe(singleMenu.date);
    expect(afterOpenModalSpy).toHaveBeenCalled();
  });

  it('Should call handleSubmit and create a new menu successfully', () => {
    const createNewMenuSpy = jest.spyOn(wrapper.instance(), 'createNewMenu');

    wrapper.setState({ selectedMeals: [{ mealId: singleMeal.id }] });
    wrapper.setProps({
      state: { updateMode: false, meals: getMealsMock(5), currentMenu: singleMenu },
    });
    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
    expect(createNewMenuSpy).toHaveBeenCalled();
  });

  it('Should call handleSubmit and update an existing menu successfully', () => {
    const updateExistingMenuSpy = jest.spyOn(wrapper.instance(), 'updateExistingMenu');

    wrapper.setState({ selectedMeals: [{ mealId: singleMeal.id }] });
    wrapper.setProps({
      state: { updateMode: true, meals: getMealsMock(5), currentMenu: singleMenu },
    });
    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
    expect(updateExistingMenuSpy).toHaveBeenCalled();
  });

  it('Should call handleSubmit and throw validation errors', () => {
    wrapper.setState({
      selectedMeals: [], name: '', startDate: '', endDate: '',
    });
    wrapper.setProps({
      state: { updateMode: true, meals: getMealsMock(5), currentMenu: singleMenu },
    });
    wrapper.find('form').simulate('submit', { preventDefault: () => { } });
    expect(wrapper.state().errors).toEqual({
      selectedMeals: 'You must select at least one meal',
      startDate: 'The date field is required',
    });
  });
});
