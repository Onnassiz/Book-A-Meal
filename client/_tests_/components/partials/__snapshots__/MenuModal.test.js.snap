// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuModal /> Should render Menu Modal page successfully 1`] = `
<div>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={1}
    contentLabel="Modal"
    isOpen={false}
    onAfterOpen={[Function]}
    onRequestClose={[MockFunction]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "WebkitOverflowScrolling": "touch",
          "background": "#ffffff",
          "border": "none",
          "borderRadius": "6px",
          "bottom": "20%",
          "left": "15%",
          "outline": "none",
          "overflow": "auto",
          "position": "absolute",
          "right": "15%",
          "top": "10%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
  >
    <div
      className="col-12"
    >
      <a
        onClick={[MockFunction]}
        style={
          Object {
            "float": "right",
          }
        }
        title="close"
      >
        <i
          className="material-icons"
          style={
            Object {
              "fontSize": 25,
            }
          }
        >
          close
        </i>
      </a>
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="box-menu"
        >
          <h3>
            Set Daily
             Menu
          </h3>
          <ShowErrors
            clientErrors={Object {}}
            serverErrors={Object {}}
          />
          <div>
            <BasicInput
              hasError={false}
              label="Start Date"
              min="2018-09-07"
              name="startDate"
              onChange={[Function]}
              type="date"
              value="2018-09-07"
            />
          </div>
          <div>
            <BasicInput
              hasError={false}
              label="Menu Name (optional)"
              min=""
              name="name"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            style={
              Object {
                "fontSize": 18,
                "marginTop": 45,
              }
            }
          >
            <span
              className="number-badge"
            >
              0
            </span>
             meals selected
          </div>
        </div>
        <div
          id="meals-checkBox"
        >
          <Checkbox
            isChecked={false}
            key="id0"
            meal={
              Object {
                "category": "dolor sit",
                "caterer": "Ben",
                "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
                "id": "id0",
                "name": "Malvaceae",
                "price": 0,
                "totalPrice": 2000,
                "units": 2,
                "userId": "84c6d80c-3d4a-4772-856c-0b3825bc537e",
              }
            }
            onChange={[Function]}
          />
          <Checkbox
            isChecked={false}
            key="id1"
            meal={
              Object {
                "category": "dolor sit",
                "caterer": "Ben",
                "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
                "id": "id1",
                "name": "Malvaceae",
                "price": 40,
                "totalPrice": 2000,
                "units": 2,
                "userId": "84c6d80c-3d4a-4772-856c-0b3825bc537e",
              }
            }
            onChange={[Function]}
          />
          <Checkbox
            isChecked={false}
            key="id2"
            meal={
              Object {
                "category": "dolor sit",
                "caterer": "Ben",
                "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
                "id": "id2",
                "name": "Malvaceae",
                "price": 80,
                "totalPrice": 2000,
                "units": 2,
                "userId": "84c6d80c-3d4a-4772-856c-0b3825bc537e",
              }
            }
            onChange={[Function]}
          />
          <Checkbox
            isChecked={false}
            key="id3"
            meal={
              Object {
                "category": "dolor sit",
                "caterer": "Ben",
                "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
                "id": "id3",
                "name": "Malvaceae",
                "price": 120,
                "totalPrice": 2000,
                "units": 2,
                "userId": "84c6d80c-3d4a-4772-856c-0b3825bc537e",
              }
            }
            onChange={[Function]}
          />
          <Checkbox
            isChecked={false}
            key="84c6d80c-3d4a-4772-856c-0b3825bc5333"
            meal={
              Object {
                "category": "dolor sit",
                "caterer": "Ben",
                "description": "Maecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat. Curabitur gravida nisi at nibh.",
                "id": "84c6d80c-3d4a-4772-856c-0b3825bc5333",
                "imageUrl": "http://thisImage.jpeg",
                "name": "Malvaceae",
                "price": 400,
                "totalPrice": 2000,
                "units": 2,
                "userId": "2794fddd-14f3-4ec9-a1dc-88ad2bc649da",
              }
            }
            onChange={[Function]}
          />
          <div
            style={
              Object {
                "textAlign": "center",
              }
            }
          />
        </div>
        <div
          id="meals"
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <hr />
          <SubmitButton
            className="button"
            isLoading={false}
            value="Update"
          />
        </div>
      </form>
    </div>
  </Modal>
</div>
`;
